stdlib/builtins.pyi

```diff
@@ -1052,6 +1052,8 @@ class filter(Iterator[_T], Generic[_T]):
     def __init__(self, __function: None, __iterable: Iterable[_T | None]) -> None: ...
     @overload
     def __init__(self, __function: Callable[[_T], Any], __iterable: Iterable[_T]) -> None: ...
+    @overload
+    def __init__(self, __function: Callable[[_S], TypeGuard[_T]], __iterable: Iterable[_S]) -> None: ...
```

Comment:
```suggestion
    @overload
    def __init__(self, __function: Callable[[_S], TypeGuard[_T]], __iterable: Iterable[_S]) -> None: ...
    @overload
    def __init__(self, __function: Callable[[_T], Any], __iterable: Iterable[_T]) -> None: ...
```

Here is a simple test script:

```python
from typing import Iterable, TypeGuard

def is_not_none(x: int | None) -> TypeGuard[int]:
    return x is not None

list_of_ints_or_none: list[int | None] = [0, None, 1, None, 2]
ints_list = list(filter(is_not_none, list_of_ints_or_none))
reveal_type(ints_list)
```

Run mypy on this script, with this patch applied, and the following is output (not the desired result):

```
test_filter.py:8: note: Revealed type is "builtins.list[Union[builtins.int, None]]"
```

However, if we change the order of the overloads, as I suggest here, then this is the output from mypy (the desired result):

```
test_filter.py:8: note: Revealed type is "builtins.list[builtins.int*]"
```

Reply:
Thanks @AlexWaygood! I was able to reproduce the result of your test.